### PR TITLE
Module: support defining const and use const variables inside of function

### PR DESCRIPTION
**Pull request by WindSoilder**
# Description

Relative: #8248 

After this pr, user can define const variable inside a module.
image

And user can export const variables, the following screenshot shows how it works (it follows https://github.com/nushell/nushell/issues/8248#issuecomment-1637442612):
image

## About the change
1. To make module support const, we need to change `parse_module_block` to support `const` keyword.
2. To suport export `const`, we need to make module tracking variables, so we add `variables` attribute to `Module`
3. During eval, the const variable may not exists in `stack`, because we don't eval `const` when we define a module, so we need to find variables which are already registered in `engine_state`

## One more thing to note about the const value.
Consider the following code
```
module foo { const b = 3; export def bar [] { $b } }
use foo bar
const b = 4;
bar
```
The result will be 3 (which is defined in module) rather than 4.  I think it's expected behavior.

It's something like [dynamic binding](https://www.gnu.org/software/emacs/manual/html_node/elisp/Dynamic-Binding-Tips.html) vs [lexical binding](https://www.gnu.org/software/emacs/manual/html_node/elisp/Lexical-Binding.html) in lisp like language, and lexical binding should be right behavior which generates more predicable result, and it doesn't introduce really subtle bugs in nushell code.

What if user want dynamic-binding?(For example: the example code returns `4`)
There is no way to do this, user should consider passing the value as argument to custom command rather than const.

## TODO
- [X] adding tests for the feature.
- [X] support export const out of module to use.

# User-Facing Changes

# Tests + Formatting

# After Submitting

### CODE REVIEW

**Comment by amtoine** [comment]
very nice, looking forward to this :blush:

**Comment by amtoine** [comment]
@WindSoilder 
i feel this should *close* #8248, right?

maybe you do not plan to support `export const ...` for now? :yum:

**Comment by WindSoilder** [comment]
Yeah, I want to support `export const` for now.
And I think the implementation code is completed currently, just need to add testing code later

File: crates/nu-cmd-lang/src/core_commands/export_const.rs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        _call: &Call,
+        _input: PipelineData,
+    ) -> Result<PipelineData, ShellError> {
+        Ok(PipelineData::empty())
+    }
+
+    fn examples(&self) -> Vec<Example> {
+        vec![Example {
+            description: "Re-export a command from another module",
+            example: r#"module spam { export def foo [] { "foo" } }
```
Comment (amtoine) [inline crates/nu-cmd-lang/src/core_commands/export_const.rs:54]: i'm not sure this example is correct, maybe that's a bad copy-paste? :yum:

**Review by amtoine — COMMENTED** [review]
for now, i've read all the non-parser files and i'm loving the tests :blush: 

if someone feels ok merging this, i'm all for it, maybe i'll come at it again later if it's not been landed by that time :relieved:

File: crates/nu-cmd-lang/src/core_commands/export_const.rs (hunk 1 above)
Comment (WindSoilder) [reply crates/nu-cmd-lang/src/core_commands/export_const.rs]: Oh, thanks for a good catch!

**Review by amtoine — APPROVED** [review]
> Oh, thanks for a good catch!
>
very nice, thanks for the fix @WindSoilder :blush: 

let's wait for the release to settle down but looking at the tests you've added, i can only approve this, thanks a lot :sparkles:

**Comment by WindSoilder** [comment]
TODO:  the constants from modules should always fold into records, like the following:

```nushell
module spam {
    export module eggs {
        export module bacon {
            export const viking = 'eats'
        }
    }
}

use spam
$spam.eggs.bacon.viking
# should print 'eats'
```

**Pull request merged**